### DISCUSSION TITLE
Pull Text from .TXT file using Colab

### DISCUSSION THREAD

**Discussion by FrankLidgad** [neonbjb/tortoise-tts#244] [Ideas]
In Colab can you add the code to a cell to pull the Text from a .TXT file as opposed to copying and pasting ?

This is available in the installed PC method but not in the Colab.

Thank you.

PS: Are there any 'Voice' repositories available other than what is in the Colab? (.WAV)

**Comment by SummerChild90** [comment]
### **Cell 1**
`!pip3 install -U scipy

!git clone https://github.com/jnordberg/tortoise-tts.git
%cd tortoise-tts
!pip3 install transformers==4.19.0
!pip3 install -r requirements.txt
!python3 setup.py install

import torch
import torchaudio
import torch.nn as nn
import torch.nn.functional as F

import IPython

from tortoise.api import TextToSpeech
from tortoise.utils.audio import load_audio, load_voice, load_voices

tts = TextToSpeech()

import os
from google.colab import files`

### **Cell 2**
`# upload at least 2 audio clips. They must be a WAV file, 6-10 seconds long.
CUSTOM_VOICE_NAME = "custom"

custom_voice_folder = f"tortoise/voices/{CUSTOM_VOICE_NAME}"
os.makedirs(custom_voice_folder)
for i, file_data in enumerate(files.upload().values()):
  with open(os.path.join(custom_voice_folder, f'{i}.wav'), 'wb') as f:
    f.write(file_data)`

### **Cell 3**
uploaded = files.upload()
for file_name, file_data in uploaded.items():
  text = file_data.decode('utf-8')
  
  # Break text into chunks of 100 characters
  text_chunks = [text[i:i+100] for i in range(0, len(text), 100)]

  # Pick a "preset mode" to determine quality. Options: {"ultra_fast", "fast" (default), "standard", "high_quality"}. See docs in api.py
  preset = "high_quality"

  # Generate speech for each chunk of text
  for i, chunk in enumerate(text_chunks):
    gen = tts.tts_with_preset(chunk, voice_samples=voice_samples, conditioning_latents=conditioning_latents, 
                              preset=preset)
    torchaudio.save(f'generated-{CUSTOM_VOICE_NAME}-{i}.wav', gen.squeeze(0).cpu(), 24000)
  
  # Combine all generated speech files into a single file
  combined = []
  for i in range(len(text_chunks)):
    file_path = f'generated-{CUSTOM_VOICE_NAME}-{i}.wav'
    combined.append(load_audio(file_path))
  combined = torch.cat(combined, dim=1)

  # Save and play the combined speech file
  torchaudio.save(f'generated-{CUSTOM_VOICE_NAME}.wav', combined, 24000)
  IPython.display.Audio(f'generated-{CUSTOM_VOICE_NAME}.wav')
